File: backend/app/services/face_service.py
import numpy as np
from uuid import uuid4
from ..db import memory_db

async def register_user(name: str, email: str, face_descriptor: list):
    user_id = str(uuid4())
    user = {
        "id": user_id,
        "name": name,
        "email": email,
        "faceDescriptor": face_descriptor
    }
    memory_db.users.append(user)
    return user

async def verify_face(face_descriptor: list):
    if not memory_db.users:
        return None
    
    input_vector = np.array(face_descriptor)
    best_match = None
    min_distance = 0.6 # Threshold
    
    for user in memory_db.users:
        stored_vector = np.array(user["faceDescriptor"])
        distance = np.linalg.norm(input_vector - stored_vector)
        
        if distance < min_distance:
            min_distance = distance
            best_match = {
                "id": user["id"],
                "name": user["name"],
                "email": user["email"]
            }
            
    return best_match

async def get_all_users():
    return [{"id": u["id"], "name": u["name"], "email": u["email"]} for u in memory_db.users]

File: backend/app/services/face_service.ts
import { users, User } from '../db/memory_db';
import { v4 as uuidv4 } from 'uuid';

export const registerUser = async (name: string, email: string, faceDescriptor: number[]): Promise<User> => {
  const newUser: User = {
    id: uuidv4(),
    name,
    email,
    faceDescriptor,
  };
  users.push(newUser);
  return newUser;
};

export const verifyFace = async (faceDescriptor: number[]): Promise<User | null> => {
  if (users.length === 0) return null;

  let bestMatch: User | null = null;
  let minDistance = 0.6; // Threshold for face recognition

  for (const user of users) {
    const distance = euclideanDistance(faceDescriptor, user.faceDescriptor);
    if (distance < minDistance) {
      minDistance = distance;
      bestMatch = user;
    }
  }

  return bestMatch;
};

export const getAllUsers = async (): Promise<User[]> => {
  return users.map(({ id, name, email }) => ({ id, name, email, faceDescriptor: [] }));
};

// Helper function to calculate distance between two descriptors
function euclideanDistance(v1: number[], v2: number[]): number {
  return Math.sqrt(v1.reduce((sum, val, i) => sum + Math.pow(val - v2[i], 2), 0));
}
